api/src/main/java/ai/djl/translate/LMSearch.java

```diff
@@ -0,0 +1,617 @@
+                nextPastHiddenStates,
+                nextOutputIds,
+                nextPastAttentionMask);
+    }
+
+    private NDArray prepareAttentionMaskOffset(NDArray inputIds, SearchConfig config) {
+        // prepare attentionMask and positionOffset
+        // Used to initialize the search
+        boolean suffixPadding = config.suffixPadding;
+        NDManager manager = inputIds.getManager();
+        int numBatch = (int) inputIds.getShape().get(0);
+        int initSeqSize = (int) inputIds.getShape().get(1);
+        NDArray attentionMask =
+                manager.ones(new Shape(1, inputIds.getShape().get(-1)), DataType.INT64)
+                        .reshape(1, -1)
+                        .repeat(0, numBatch);
+
+        // Linear search from left to find the first position that's not padTokenId.
+        long[][] offset = new long[numBatch][1];
+        for (int i = 0; i < numBatch; i++) {
+            long[] aSequence = inputIds.get("{},:", i).toLongArray();
+            int idx = 0;
+            while (idx < initSeqSize) {
+                if (suffixPadding && aSequence[idx] == config.padTokenId
+                        || !suffixPadding && aSequence[idx] != config.padTokenId) {
+                    break;
+                }
+                idx++;
+            }
+            attentionMask.set(
+                    new NDIndex(
+                            "{},{}:{}",
+                            i,
+                            suffixPadding ? idx : 0,
+                            suffixPadding ? initSeqSize : idx),
+                    0);
+            if (!suffixPadding) {
+                offset[i][0] = idx;
+            }
+        }
+        positionOffset = manager.create(offset);
+        return attentionMask;
+    }
+
+    private NDList prepareInput(
+            NDArray inputIds, NDArray attentionMask, long pastSeqLength, int repeat) {
+        // Pack the model input
+        NDArray positionIds =
+                inputIds.getManager()
+                        .arange(
+                                pastSeqLength,
+                                pastSeqLength + inputIds.getShape().get(-1),
+                                1,
+                                DataType.INT64)
+                        .expandDims(0)
+                        .repeat(0, inputIds.getShape().get(0));
+
+        NDArray positionIdsShifted = positionIds.subi(positionOffset.repeat(0, repeat));
+        positionIds = positionIdsShifted.maximum(positionIdsShifted.zerosLike());
+
+        return new NDList(inputIds, positionIds, attentionMask);
+    }
+
+    public NDArray forward(NDArray inputIds) {
+        switch (searchName) {
+            case "greedy":
+                return greedySearch(inputIds);
+            case "beam":
+                return beamSearch(inputIds);
+            case "contrastive":
+                return contrastiveSearch(inputIds);
+            default:
+                return greedySearch(inputIds);
```

Comment:
I would consider breaking up `LMSearch` into an abstract `LMSearch` class and four specific search blocks that extend it. You can still share helper code like `prepareAttentionMaskOffset` by putting it in the parent `LMSearch`.

Otherwise, you should probably throw an Exception for the default behavior rather than just use greedySearch. When in doubt when something is probably going wrong, it is better to make sure the user is aware and can fix it rather than risk doing the wrong thing to compensate for the user's mistake